presto-hive/src/main/java/com/facebook/presto/hive/authentication/CachingKerberosHadoopAuthentication.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.authentication;
+
+import com.facebook.presto.hive.HiveErrorCode;
+import com.facebook.presto.spi.PrestoException;
+import org.apache.hadoop.security.UserGroupInformation;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.security.auth.Subject;
+import javax.security.auth.kerberos.KerberosTicket;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+import static com.facebook.presto.hive.authentication.KerberosTicketUtils.getTgt;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class CachingKerberosHadoopAuthentication
+        implements HadoopAuthentication
+{
+    private final KerberosHadoopAuthentication delegate;
+
+    private final Object lock = new Object();
+    @GuardedBy("lock")
+    private UserGroupInformation userGroupInformation = null;
+    @GuardedBy("lock")
+    private long nextRefreshTime = Long.MIN_VALUE;
+
+    public CachingKerberosHadoopAuthentication(KerberosHadoopAuthentication delegate)
+    {
+        this.delegate = requireNonNull(delegate, "hadoopAuthentication is null");
+    }
+
+    @Override
+    public UserGroupInformation getUserGroupInformation()
+    {
+        if (refreshIsNeeded()) {
+            synchronized (lock) {
+                if (refreshIsNeeded()) {
+                    refreshUgi();
+                }
+            }
+        }
+        return userGroupInformation;
+    }
+
+    private void refreshUgi()
+    {
+        this.userGroupInformation = delegate.getUserGroupInformation();
+        this.nextRefreshTime = calculateNextRefreshTime(this.userGroupInformation);
+    }
+
+    private boolean refreshIsNeeded()
+    {
+        return nextRefreshTime < System.currentTimeMillis() || userGroupInformation == null;
+    }
+
+    private long calculateNextRefreshTime(UserGroupInformation userGroupInformation)
+    {
+        Subject subject = getSubject(userGroupInformation);
+        checkState(subject != null, "subject must be present in kerberos based UGI");
+        KerberosTicket tgtTicket = getTgt(subject);
+        return KerberosTicketUtils.getRefreshTime(tgtTicket);
+    }
+
+    private Subject getSubject(UserGroupInformation userGroupInformation)
+    {
+        try {
+            Method getSubjectMethod = UserGroupInformation.class.getDeclaredMethod("getSubject");
```

Comment:
@electrum @dain 

I'm going to add the next Shim to the `presto-hadoop-apache1`, `presto-hadoop-apache2` and `presto-hadoop-cdh4` libraries and create the PR's.

``` java
public final class UserGroupInformationShim
{
    private UserGroupInformationShim() {}

    public static Subject getSubject(UserGroupInformation userGroupInformation)
    {
        return userGroupInformation.getSubject();
    }

    public static UserGroupInformation createUserGroupInformationForSubject(Subject subject)
    {
        if (subject == null) {
            throw new NullPointerException("subject is null");
        }
        Set<KerberosPrincipal> kerberosPrincipals = subject.getPrincipals(KerberosPrincipal.class);
        if (kerberosPrincipals.isEmpty()) {
            throw new IllegalArgumentException("subject must contain a KerberosPrincipal");
        }
        if (kerberosPrincipals.size() != 1) {
            throw new IllegalArgumentException("subject must contain only a single KerberosPrincipal");
        }

        KerberosPrincipal principal = kerberosPrincipals.iterator().next();
        User user = new User(principal.getName(), KERBEROS, null);
        subject.getPrincipals().add(user);
        UserGroupInformation userGroupInformation = new UserGroupInformation(subject);
        userGroupInformation.setAuthenticationMethod(KERBEROS);
        return userGroupInformation;
    }
}
```

Could you please confirm that this implementation looks good to you before i started to create those PR's.